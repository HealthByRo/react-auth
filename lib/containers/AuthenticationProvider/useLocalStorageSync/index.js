"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = useLocalStorageSync;

var _react = require("react");

var _setAuthDataInStorage = _interopRequireDefault(require("../utils/setAuthDataInStorage"));

var _storeLastUserToken = _interopRequireDefault(require("../utils/storeLastUserToken"));

var _generateLastUserTokenKey = _interopRequireDefault(require("../utils/generateLastUserTokenKey"));

var _removeAuthDataFromStorage = _interopRequireDefault(require("../utils/removeAuthDataFromStorage"));

function useLocalStorageSync(tokenData, userData) {
  return (0, _react.useEffect)(function () {
    (0, _setAuthDataInStorage["default"])({
      tokenData: tokenData
    }); // if (!tokenData) {
    //   removeAuthDataFromStorage();
    // }

    if (tokenData && userData) {
      var lastUserTokenKey = (0, _generateLastUserTokenKey["default"])(userData.email);
      (0, _storeLastUserToken["default"])(lastUserTokenKey, tokenData.key);
    }
  }, [tokenData, userData]);
}