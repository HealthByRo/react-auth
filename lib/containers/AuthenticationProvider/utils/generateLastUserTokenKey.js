"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = generateLastUserTokenKey;

var _md = _interopRequireDefault(require("md5"));

function generateLastUserTokenKey(rawKey) {
  return (0, _md["default"])(rawKey.toLowerCase().trim());
}