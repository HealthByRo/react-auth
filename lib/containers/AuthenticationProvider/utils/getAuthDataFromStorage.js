"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = getAuthDataFromStorage;

var _jsBase = require("js-base64");

var _constants = require("../constants");

var _getItemFromStorage = _interopRequireDefault(require("./getItemFromStorage"));

function getAuthDataFromStorage() {
  var localStorageItem = (0, _getItemFromStorage["default"])(_constants.AUTH_KEY);

  if (localStorageItem) {
    try {
      var decodedLocalStorageItem = _jsBase.Base64.decode(localStorageItem);

      return JSON.parse(decodedLocalStorageItem);
    } catch (e) {
      return null;
    }
  }

  return localStorageItem;
}