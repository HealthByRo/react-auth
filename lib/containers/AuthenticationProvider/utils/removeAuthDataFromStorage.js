"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = removeAuthDataFromStorage;

var _constants = require("../constants");

function removeAuthDataFromStorage() {
  try {
    localStorage.removeItem(_constants.AUTH_KEY);
    return true;
  } catch (e) {
    return false;
  }
}