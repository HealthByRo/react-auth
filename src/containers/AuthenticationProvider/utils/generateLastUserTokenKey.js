import md5 from 'md5';

export default function generateLastUserTokenKey(rawKey) {
  return md5(rawKey.toLowerCase().trim());
}
