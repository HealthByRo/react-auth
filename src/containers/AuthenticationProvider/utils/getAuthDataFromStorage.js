import { Base64 } from 'js-base64';
import { AUTH_KEY } from '../constants';
import getItemFromStorage from './getItemFromStorage';

export default function getAuthDataFromStorage() {
  const localStorageItem = getItemFromStorage(AUTH_KEY);

  if (localStorageItem) {
    try {
      const decodedLocalStorageItem = Base64.decode(localStorageItem);

      return JSON.parse(decodedLocalStorageItem);
    } catch (e) {
      return null;
    }
  }

  return localStorageItem;
}
