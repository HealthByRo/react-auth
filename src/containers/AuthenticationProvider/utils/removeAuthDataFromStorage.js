import { AUTH_KEY } from '../constants';

export default function removeAuthDataFromStorage() {
  try {
    localStorage.removeItem(AUTH_KEY);
    return true;
  } catch (e) {
    return false;
  }
}
