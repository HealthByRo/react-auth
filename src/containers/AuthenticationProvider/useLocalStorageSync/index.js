import { useEffect } from 'react';
import setAuthDataInStorage from '../utils/setAuthDataInStorage';
import storeLastUserToken from '../utils/storeLastUserToken';
import generateLastUserTokenKey from '../utils/generateLastUserTokenKey';
import removeAuthDataFromStorage from '../utils/removeAuthDataFromStorage';

export default function useLocalStorageSync(tokenData, userData) {
  return useEffect(() => {
    setAuthDataInStorage({ tokenData });

    if (!tokenData) {
      removeAuthDataFromStorage();
    }

    if (tokenData && userData) {
      const lastUserTokenKey = generateLastUserTokenKey(userData.email);

      storeLastUserToken(lastUserTokenKey, tokenData.key);
    }
  }, [tokenData, userData]);
}
