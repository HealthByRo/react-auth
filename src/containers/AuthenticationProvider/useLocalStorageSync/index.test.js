import React, { useState } from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';
import useLocalStorageSync from '.';
import getAuthDataFromStorage from '../utils/getAuthDataFromStorage';
import removeAuthDataFromStorage from '../utils/removeAuthDataFromStorage';
import generateLastUserTokenKey from '../utils/generateLastUserTokenKey';

jest.mock('../../../api', () => ({
  removeAuthorizationTokenInHeaders: jest.fn(),
  setAuthorizationTokenInHeaders: jest.fn(),
}));

let mockTokenData;
let mockUserData;
let mockTokenKey;

function TestComponent() {
  const [tokenData, setTokenData] = useState();
  const [userData, setUserData] = useState();

  useLocalStorageSync(tokenData, userData);

  return (
    <>
      <button
        type="button"
        onClick={() => setUserData(mockUserData)}
      >
        Set user data
      </button>

      <button
        type="button"
        onClick={() => setTokenData(mockTokenData)}
      >
        Set token data
      </button>

      <button
        type="button"
        onClick={() => {
          setTokenData(mockTokenData);
          setUserData(mockUserData);
        }}
      >
        Set both: user and token data
      </button>
    </>
  );
}

describe('processAuthResponse', () => {
  let container;

  beforeEach(() => {
    container = render(<TestComponent />);
  });

  afterEach(() => {
    removeAuthDataFromStorage();
  });

  describe('when clicks on "Set token data" button which set mocked token data', () => {
    beforeEach(() => {
      mockTokenData = {
        key: 'TEST_TOKEN',
        status: 'TEST_STATUS',
        expiredAt: 'TEST_DATE_TIME',
      };

      const button = container.queryByText('Set token data');

      fireEvent.click(button);
    });

    it('auth data saved in localStorage is equal to mockTokenData', () => {
      expect(getAuthDataFromStorage().tokenData).toEqual(mockTokenData);
    });

    it('last tokens saved in auth data in localStorage are defined (is object which contains hashed email as a key and toke.key as value)', () => {
      expect(getAuthDataFromStorage().lastTokens).toBeUndefined();
    });

    it('auth data saved in localStorage match snapshot', () => {
      expect(getAuthDataFromStorage()).toMatchSnapshot();
    });

    describe('when clicks on "Set user data" button which set mocked user data', () => {
      beforeEach(() => {
        mockUserData = {
          id: 'TEST_USER_ID',
          email: 'TEST_USER_EMAIL',
        };

        const button = container.queryByText('Set user data');

        fireEvent.click(button);
      });

      it('last tokens saved in auth data in localStorage are not defined', () => {
        expect(getAuthDataFromStorage().lastTokens).toEqual({
          [generateLastUserTokenKey(mockUserData.email)]: mockTokenData.key,
        });
      });

      it('auth data saved in localStorage match snapshot', () => {
        expect(getAuthDataFromStorage()).toMatchSnapshot();
      });

      describe('when clicks on "Set both: user and token data" button which set mocked token data and user data at the same time', () => {
        let oldMockTokenData;
        let oldMockUserData;
        mockTokenData = {
          key: 'TEST_TOKEN_UPDATED',
          status: 'TEST_STATUS_UPDATED',
          expiredAt: 'TEST_DATE_TIME_UPDATED',
        };
        mockUserData = {
          id: 'TEST_USER_ID_UPDATED',
          email: 'TEST_USER_EMAIL_UPDATED',
        };

        beforeEach(() => {
          oldMockTokenData = mockTokenData;
          oldMockUserData = mockUserData;
          mockTokenKey = 'TEST_TOKEN_UPDATED';
          mockTokenData = {
            key: mockTokenKey,
            status: 'TEST_STATUS_UPDATED',
            expiredAt: 'TEST_DATE_TIME_UPDATED',
          };
          mockUserData = {
            id: 'TEST_USER_ID_UPDATED',
            email: 'TEST_USER_EMAIL_UPDATED',
          };

          const button = container.queryByText('Set both: user and token data');

          fireEvent.click(button);
        });

        it('auth data saved in localStorage is equal to updated mockTokenData', () => {
          expect(getAuthDataFromStorage().tokenData).toEqual(mockTokenData);
        });

        it('last tokens saved in auth data in localStorage are defined and contains two entries', () => {
          expect(getAuthDataFromStorage().lastTokens).toEqual({
            [generateLastUserTokenKey(oldMockUserData.email)]: oldMockTokenData.key,
            [generateLastUserTokenKey(mockUserData.email)]: mockTokenData.key,
          });
        });

        it('auth data saved in localStorage match snapshot', () => {
          expect(getAuthDataFromStorage()).toMatchSnapshot();
        });
      });

      describe('when clicks on "Set token data" button which set null (it means user logout)', () => {
        beforeEach(() => {
          mockTokenData = null;

          const button = container.queryByText('Set token data');

          fireEvent.click(button);
        });

        it('auth data in localStorage is undefined', () => {
          expect(getAuthDataFromStorage()).toBeUndefined();
        });

        it('auth data saved in localStorage match snapshot', () => {
          expect(getAuthDataFromStorage()).toMatchSnapshot();
        });
      });
    });
  });

  describe('when clicks on "Set user data" button which set mocked user data (setting only user data does not save anything in local storage', () => {
    beforeEach(() => {
      mockUserData = {
        id: 'TEST_USER_ID',
        email: 'TEST_USER_EMAIL',
      };

      const button = container.queryByText('Set user data');

      fireEvent.click(button);
    });

    it('auth data saved in localStorage is empty undefined', () => {
      expect(getAuthDataFromStorage()).toBeUndefined();
    });
  });
});
