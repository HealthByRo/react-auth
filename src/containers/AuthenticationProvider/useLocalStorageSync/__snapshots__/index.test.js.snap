// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processAuthResponse when clicks on "Set token data" button which set mocked token data auth data saved in localStorage match snapshot 1`] = `
Object {
  "tokenData": Object {
    "expiredAt": "TEST_DATE_TIME",
    "key": "TEST_TOKEN",
    "status": "TEST_STATUS",
  },
}
`;

exports[`processAuthResponse when clicks on "Set token data" button which set mocked token data when clicks on "Set user data" button which set mocked user data auth data saved in localStorage match snapshot 1`] = `
Object {
  "lastTokens": Object {
    "227abe90793934fc2588650db74319b7": "TEST_TOKEN",
  },
  "tokenData": Object {
    "expiredAt": "TEST_DATE_TIME",
    "key": "TEST_TOKEN",
    "status": "TEST_STATUS",
  },
}
`;

exports[`processAuthResponse when clicks on "Set token data" button which set mocked token data when clicks on "Set user data" button which set mocked user data when clicks on "Set both: user and token data" button which set mocked token data and user data at the same time auth data saved in localStorage match snapshot 1`] = `
Object {
  "lastTokens": Object {
    "227abe90793934fc2588650db74319b7": "TEST_TOKEN",
    "f7e5d98e44426ec62dcc7119a9aa61fb": "TEST_TOKEN_UPDATED",
  },
  "tokenData": Object {
    "expiredAt": "TEST_DATE_TIME_UPDATED",
    "key": "TEST_TOKEN_UPDATED",
    "status": "TEST_STATUS_UPDATED",
  },
}
`;

exports[`processAuthResponse when clicks on "Set token data" button which set mocked token data when clicks on "Set user data" button which set mocked user data when clicks on "Set token data" button which set null (it means user logout) auth data saved in localStorage match snapshot 1`] = `undefined`;
